compiler/z/codegen/ControlFlowEvaluator.cpp

```diff
@@ -2669,18 +2669,26 @@ OMR::Z::TreeEvaluator::dselectEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    TR::Register *resultReg = cg->gprClobberEvaluate(trueValueNode);
    TR::Register *conditionReg = cg->evaluate(conditionNode);
    TR::Register *falseValReg = cg->evaluate(falseValueNode);
-   if (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_S390_Z13) && node->getOpCode().isDouble())
+   if ((cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_S390_Z13) && node->getOpCode().isDouble()) || (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_S390_Z14) && node->getOpCode().isFloat()))
       {
       TR::Register *vectorSelReg = cg->allocateRegister(TR_VRF);
       TR::Register *tempReg = cg->allocateRegister(TR_FPR);
       TR::Register *vzeroReg = cg->allocateRegister(TR_VRF);
-      // Convert 32 Bit register to 64 Bit (Comparison Child of the select node is 32 bit)
-      generateRRInstruction(cg, TR::InstOpCode::LLGFR, node, conditionReg, conditionReg);
+      
+      // Convert 32 Bit register to 64 Bit for Doubles (Comparison Child of the select node is 32 bit)
+      if (node->getOpCode().isDouble())
+       {
+        generateRRInstruction(cg, TR::InstOpCode::LLGFR, node, conditionReg, conditionReg);
+       }else  
+       {
+        generateRSInstruction(cg, TR::InstOpCode::SLLG, node, conditionReg, 32);
+       }
+
       // convert to floating point
       generateRRInstruction(cg, TR::InstOpCode::LDGR, node, tempReg, conditionReg);
       // generate compare with zero
       generateVRIaInstruction(cg, TR::InstOpCode::VGBM, node, vzeroReg, 0, 0);
-      generateVRRcInstruction(cg, TR::InstOpCode::VFCE, node, vectorSelReg, tempReg, vzeroReg, 1, 0, 3);
+      generateVRRcInstruction(cg, TR::InstOpCode::VFCE, node, vectorSelReg, tempReg, vzeroReg, 0, 0x8, getVectorElementSizeMask(node->getSize()));
```

Comment:
Please comment the purpose behind mask values here.

Reply:
For the VFCE instruction, the following mask values have been used to generate vector instructions for both doubles and floats.

M4 - Floating-point-format control = **getVectorElementSizeMask(node->getSize());** returns 2 for short format floats and 3 for long format doubles accordingly, depending on the node.

M5 - Single-Element-Control = **0x8**, which sets the bit 0 to one, which would control the operation to take place only on the zero-indexed element in the vector.

M6 - Condition Code Set = 0, the Condition Code is not set and remains unchanged.


Binary Encoding for Double:

```
0x2aa2ea3f8a0] b9 16 00 00                                       LLGFR   GPR0,GPR0
     0x3ff8990013a 00000042 [     0x2aa2ea3f950] b3 c1 00 20                                       LDGR    FPR2,GPR0
     0x3ff8990013e 00000046 [     0x2aa2ea3fa00] e7 00 00 00 08 44                                 VGBM    VRF16,0x0
     0x3ff89900144 0000004c [     0x2aa2ea3fad0] e7 02 00 08 3a e8                                 WFCEDB  VRF16,VRF2,VRF16
```

Binary encoding for Float:

```
0x2aa3be9ccf0] eb 00 00 20 00 0d                                 SLLG    GPR0,32
     0x3ff88280140 00000048 [     0x2aa3be9cdb0] b3 c1 00 20                                       LDGR    FPR2,GPR0
     0x3ff88280144 0000004c [     0x2aa3be9ce60] e7 00 00 00 08 44                                 VGBM    VRF16,0x0
     0x3ff8828014a 00000052 [     0x2aa3be9cf30] e7 02 00 08 2a e8                                 WFCEDB  VRF16,VRF2,VRF16
```